Move files/ folder between folders.

pablocarmu opened the pull request:
<!--
Thanks for submitting a change to ownCloud!

To make it possible for us to get your change reviewed and merged please fill out below information carefully.

Please note that any kind of change first has to be submitted to the master branch which holds the next major version of ownCloud.
-->

## Description
<!--- Describe your changes in detail -->
This PR implements the movement of files and folders between folders.
It also implements the first _drag and drop_ support.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
With this implementation, you can select and multi-select files and folders and move it through the directory tree.

_the multi-select feature is only available using drag and drop for now_

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/owncloud/ios-app/blob/master/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

_____

BUGS & IMPROVEMENTS

- [X] (1) No option to confirm in empty folders https://github.com/owncloud/ios-app/pull/110#issuecomment-424236341 [FIXED]
- [X] (2) Move to root folder https://github.com/owncloud/ios-app/pull/110#issuecomment-424248931  [FIXED]
- [x] (3) Offline mode https://github.com/owncloud/ios-app/pull/110#issuecomment-424640397 [MOVED TO OFFLINE]
- [X] (4) Duplications remove thumbnails https://github.com/owncloud/ios-app/pull/110#issuecomment-424642085 [WONT FIX HERE]
- [X] (5) Error message when target folder does not exist https://github.com/owncloud/ios-app/pull/110#issuecomment-424668544 [FIXED in SDK]

Commit:
- First implementation of drag & drop with multiple files support (WIP).
- First implementation of the move feature through a new screen (WIP).

Commit:
- Coded the delegate methods for drag & drop support.
- Removed unneded code for drag & drop in ClientItemCell.

Commit:
- Coded hidden footer in ClientDirectoryPickerViewController to avoid extra cell separators

Commit:
- Localize 'Move Here' message

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/owncloud/ios-app?pullRequest=110) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/owncloud/ios-app?pullRequest=110) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: pablocarmu<br/>:x: felix-schwarz<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/owncloud/ios-app?pullRequest=110) it.</sub>

Commit:
- Move feature now it's part of the More view.
- Duplicate feature it's now part of the More view

Commit:
- Ban drop items over items that are not folders

pablocarmu commented:
This PR it is now updated with the today's latest version of the SDK's file management branch (commit [f8aae3a1b5f8987276afac068bcaa7998e5e5e90](https://github.com/owncloud/ios-sdk/pull/24/commits/f8aae3a1b5f8987276afac068bcaa7998e5e5e90)) so @jesmrec from my point of view you can start testing it.

jesmrec commented:
Let's QA this

jesmrec commented:
### (1)

1. Select `Move`
2. Select an empty folder as target

Current: No option to confirm
Expected: Option to confirm

iPad v11

pablocarmu committed:
- Bring to the front of the view the "move here" button to fix when there is no elements in a folder.

jesmrec commented:
### (2)

1. Move content placed into another folder 8not in root) to root folder

Current: only posible in an empty cell. Not easy from UX pov
Expected: moved to root over any file. The point is, if there is no files in the folder...

pablocarmu committed:
- Now if you drop one item inside another that is a file, it takes the file's folder path and will move the file to it.

jesmrec commented:
### (3) 

SDK related?: 

https://github.com/owncloud/ios-sdk/issues/25#issuecomment-424640163

jesmrec commented:
### (4) 

About duplications:

Duplicating a file, adds `copy` as suffix in the filename, including the extension. 

`File.txt` ->  `File.txt copy`

so, the thumbnail or preview is lost.

pablocarmu committed:
- Removed duplicate feature from this PR.

pablocarmu commented:
About https://github.com/owncloud/ios-app/pull/110#issuecomment-424642085

I've split this feature into another pull request to be able to test and review it in a more isolated way.
Should be fixed here [duplicate files and folders](https://github.com/owncloud/ios-app/pull/119)

jesmrec commented:
### (5)

(SDK involved, maybe)

1. Move folder A into folder B
2. Before confirming the action, remove the B folder in web UI
3. Confirm

Current: Message: `A not found. The target directory /B/A does not seem to exist`.

Expected: Message about the non-existence of the target folder (B), instead the source folder.

Tested in iPadAir2 v11

felix-schwarz commented:
@jesmrec (5) was indeed an issue in the SDK. The fix for it is in the next `feature/filemgmt` commit.

jesmrec commented:
all OK from my side.

pablocarmu committed:
- Removed redundant Optional initializers.

felix-schwarz commented on ownCloud/Client/Actions/ClientDirectoryPickerViewController.swift:
```diff
@@ -0,0 +1,97 @@
+	init(core inCore: OCCore, query inQuery: OCQuery, completion: @escaping (OCItem) -> Void) {
+		selectButton = ThemeButton()
+		self.completion = completion
+		super.init(core: inCore, query: inQuery)
```
Since this is a directory picker, it would be a nice touch to add a filter to the query, so it only provides directories. Right now, the picker also shows files and allows selecting them.

May also be worth considering creating the OCQuery in the VC, based on a path that's provided instead.

pablocarmu committed:
- SDK Updated

pablocarmu committed:
- Removed unused MobileCoreServices from OCItem+Extension.
- Removed unused 'canHandle...' function from OCItem+Extension.
- Fix Move here typo.

pablocarmu commented on ownCloud/Client/Actions/ClientDirectoryPickerViewController.swift:
```diff
@@ -0,0 +1,97 @@
+		])
+
+		cancelBarButton = UIBarButtonItem(barButtonSystemItem: .cancel, target: self, action: #selector(cancelBarButtonPressed))
+		navigationItem.rightBarButtonItem = cancelBarButton
```
I can do it but then, what happens with the go-back arrow?
It is ok when you are in the first folder but when you go to another one, the go-back arrow shows up. Here it's an example of an iPad.

![image-1](https://user-images.githubusercontent.com/8929738/46949308-41513600-d081-11e8-96d6-bd804fb73987.jpg)

felix-schwarz replied:
You're right, of course. Let's go for the `Move here` button solution (+ taking care of contentInsets) then.

pablocarmu replied:
I didn't noticed these things before, thanks!
I followed your suggestion and here it's the result:

![screen shot 2018-10-17 at 10 15 08](https://user-images.githubusercontent.com/8929738/47072222-e8ef7500-d1f5-11e8-9168-163d7f974407.png)

felix-schwarz replied:
Thanks! Looks great!

pablocarmu replied:
> whether actions (via row actions, action dots and drag and drop) should be available inside ClientDirectoryPickerViewController. Is it more useful or more confusing?

I think this picker view is only going to be prompted in this two actions:
 1. Select the folder in which you want to move files/folders
 2. Select the folder you want to be the default folder for automatic uploads 

There could be more use cases for this screen in the future but I tend to think it as an isolated view and feature. I'll prefer to keep it as simple as possible and focused in the _pick a folder_ task. 

While developing this feature something else came to my mind. This class is a `ClientQueryViewController` subclass, but eventually, `ClientQueryViewController` will gain more features and this would lead to overriding methods to make the picker do nothing on some actions or use cases.

I think we could make a base class with only a list of files and the sort bar/search bar and then through subclassing, we can make the `ClientQueryViewController` with the multiselection feature, drag and drop support,  3 dots in a cell for actions. And make other class like `DirectoryPickerViewController` with a filter inside for only showing folders and no extra thing but the `Move here` button.

> whether we should add a validation block property to ClientDirectoryPickerViewController, which can be used to "validate" the currently viewed directory, so that f.ex. "Move here" is disabled when the user would otherwise try to move a folder into itself - or to the location it already resides in. Right now, if a user does this, an error will pop up after tapping "Move here".

I just had a quick look at what other apps do (Dropbox) and they allow you to select the current folder as the destination for the move action. Ther prompt the user with an alert saying _Can't move to the same folder_. We can follow this way or we can compare if the `query.rootItem` is the same that the item path we're trying to move and if this is true, disable the _move here_ button

felix-schwarz replied:
@pablocarmu 

> I think we could make a base class with only a list of files and the sort bar/search bar and then through subclassing, we can make the `ClientQueryViewController` with the multiselection feature, drag and drop support, 3 dots in a cell for actions. And make other class like `DirectoryPickerViewController` with a filter inside for only showing folders and no extra thing but the `Move here` button.

I think this is reasonable, but may reach its limits if f.ex. two subclasses should share one additional feature (only that one, not sharing any others). In that case, we'd have to subclass the base class and then subclass that subclass again, ending up with a lot of classes.

An alternative could be to introduce modes to `ClientQueryViewController`, so that code can be shared between modes very selectively. Also, `DirectoryPickerViewController` would no longer be needed and just become a mode of `ClientQueryViewController`. The code would then stay in one place, which might make it easier to navigate long-term, too.

> I just had a quick look at what other apps do (Dropbox) and they allow you to select the current folder as the destination for the move action. Ther prompt the user with an alert saying _Can't move to the same folder_. We can follow this way or we can compare if the `query.rootItem` is the same that the item path we're trying to move and if this is true, disable the _move here_ button

Thanks for checking! It all boils down to "prevent impossible actions from the start" vs. "allow triggering impossible actions and then explain why they aren't possible".

The benefit of the latter is that nobody will end up thinking that "move doesn't work (because the move button is disabled)".

pablocarmu committed:
- Coded new filter to only show folders in the directory picker.
- Made the query inside the directory picker so now the init argument is
the path.

pablocarmu committed:
- Made the bottom inset of the directory picker table view to match the
heigth of the picker bottom button.
- Made editable the name of the directory picker bottom button.

pablocarmu committed:
Merge branch 'master' into feature/move

pablocarmu committed:
- Replaced the bottom move button with a UIToolbar and a UIBarButton.

jesmrec commented:
the target of the current issue was creating the move feature available in the file list. The ideas you mentioned to move some actions to Folder Picker are very interesting, but i think they should be moved to a separate issue in order to move forward the current one, that was opened almost two months ago. We should work in an incremental way and avoid performing more and more and more improvements inside the same PR.

@pablocarmu @felix-schwarz

felix-schwarz committed:
Merge branch 'master' into feature/move

felix-schwarz commented:
Please move forward with merging this PR @jesmrec. Following your suggestion, work on `ClientQueryViewController` will take place in a new PR.
